File: go/fiber/internal/modules/transaction/service.go
package transaction

import (
	r "tcc/internal/common/enum"
	"tcc/internal/common/errors"
	"tcc/internal/common/helper"
	"tcc/internal/common/messages"
	"tcc/internal/database"
	repository "tcc/internal/modules/account"
	account "tcc/internal/modules/account/entity"
	"tcc/internal/modules/transaction/dto"
	"tcc/internal/modules/transaction/entity"
	"tcc/internal/modules/transaction/enum"
)

type TransactionService struct{
	repository struct{
		account repository.AccountRepository
		transaction TransactionsRepository
	}
}

func (s *TransactionService) Create(data dto.CreateTransactionDto, idClient string) (*entity.TransactionEntity, *errors.BaseError) {
	
	var tipo = data.Tipo
	var originAccount *account.AccountEntity = nil

	if data.IDContaDestino == data.IDContaOrigem {
		return nil, errors.BadRequestError(messages.ErrorMessages.Account.BadRequest.SameAccount)
	}

	
	// Verificar se contas existem
	if data.IDContaOrigem != nil {
		originAccount, _ = s.repository.account.FindById(*data.IDContaOrigem, false)
	
		if originAccount == nil {
			return nil, errors.NotFoundError(messages.ErrorMessages.Account.NotFoundOrigin)
		}

		if originAccount.IDCliente.String() != idClient {
			return nil, errors.UnauthorizedError(messages.ErrorMessages.Unauthorized)
		}
	}

	if data.IDContaDestino != nil {
		targetAccount, _ := s.repository.account.FindById(*data.IDContaDestino, false)

		if targetAccount == nil {
			return nil, errors.NotFoundError(messages.ErrorMessages.Account.NotFoundTarget)
		}
	}


	if (tipo == enum.TRANSFER || tipo == enum.WITHDRAWAL) && (originAccount != nil && *originAccount.Saldo < data.Valor) {
		return nil, errors.BadRequestError(messages.ErrorMessages.Account.BadRequest.BalanceNotEnough)
	}


	// Iniciar transação
	tx, err := database.StartTransaction(s.repository.transaction.db)
	if err != nil {
		return nil, errors.InternalServerError(messages.ErrorMessages.InternalServer)
	}

	newTransaction, err := s.repository.transaction.Create(tx, ICreateTransactionData{
		Tipo: tipo,
		Valor: data.Valor,
		IDContaOrigem: data.IDContaOrigem,
		IDContaDestino: data.IDContaDestino,
	})

	if err != nil {
		database.EndTransaction(tx, err)
		return nil, errors.InternalServerError(messages.ErrorMessages.InternalServer)
	}

	switch tipo {
	case enum.DEPOSIT:
		err = s.repository.account.AddBalance(tx, *data.IDContaDestino, data.Valor)
	case enum.TRANSFER:
		err = s.repository.account.RemoveBalance(tx, *data.IDContaOrigem, data.Valor)
		if err == nil {
			err = s.repository.account.AddBalance(tx, *data.IDContaDestino, data.Valor)
		}
	case enum.WITHDRAWAL:
		err = s.repository.account.RemoveBalance(tx, *data.IDContaOrigem, data.Valor)
	}



	success := database.EndTransaction(tx, err)

	if !success{
		return nil, errors.InternalServerError(messages.ErrorMessages.InternalServer)
	}

	return newTransaction, nil
}

func (s *TransactionService) FindAll(idAccount string, idClient string, role r.RoleEnum) ([]*entity.TransactionEntity, *errors.BaseError) {
	permission := helper.HasPermission(role, r.MANAGER)

	account, _ := s.repository.account.FindById(idAccount, false)

	if account == nil {
		return nil, errors.NotFoundError(messages.ErrorMessages.Account.NotFound)
	}

	if account.IDCliente.String() != idClient && !permission {
		return nil, errors.UnauthorizedError(messages.ErrorMessages.Unauthorized)
	}

	transactions, _ := s.repository.transaction.FindAll(idAccount)

	if transactions == nil {
		return []*entity.TransactionEntity{}, nil
	}

	return transactions, nil
}

File: rust/actix/src/modules/transaction/service.rs
use crate::common::enums::role::RoleEnum;
use crate::modules::transaction::dto::CreateTransactionDto;
use crate::modules::transaction::entity::TransactionEntity;
use crate::modules::account::entity::AccountEntity;
use crate::common::errors::{bad_request::bad_request_error, base::BaseError, not_found::not_found_error, unauthorized::unauthorized_error};
use crate::common::messages::error::ErrorMessages;
use crate::modules::account::repository::AccountRepository;
use crate::modules::transaction::repository::TransactionRepository;
use crate::modules::transaction::enums::TransactionType;
use crate::modules::transaction::repository::ICreateTransactionData;
use crate::database;
use std::convert::From;
use bigdecimal::BigDecimal;
use crate::common::errors::internal_server::internal_server_error;
use crate::common::helper::permission;
use uuid::Uuid;

pub struct TransactionService {
}

impl TransactionService {
    pub async fn create(
        data: CreateTransactionDto,
        id_client: String,
    ) -> Result<TransactionEntity, BaseError> {
        let tipo = data.tipo;
        let mut origin_account: Option<AccountEntity> = None;

        if data.id_conta_destino == data.id_conta_origem {
            return Err(bad_request_error(ErrorMessages::ACCOUNT_BAD_REQUEST_SAME_ACCOUNT, None));
        }

        // Verificar se contas existem
        if let Some(id_conta_origem) = data.id_conta_origem {
            origin_account = AccountRepository::find_by_id(&id_conta_origem.to_string(), false)
                .await
                .map_err(|_| internal_server_error(ErrorMessages::INTERNAL_SERVER, None))?;

            if origin_account.is_none() {
                return Err(not_found_error(ErrorMessages::ACCOUNT_NOT_FOUND_ORIGIN, None));
            }

            if origin_account.as_ref().unwrap().id_cliente.unwrap().to_string() != id_client {
                return Err(unauthorized_error(ErrorMessages::UNAUTHORIZED, None));
            }
        }

        if let Some(id_conta_destino) = data.id_conta_destino {
            let target_account = AccountRepository::find_by_id(&id_conta_destino.to_string(), false)
                .await
                .map_err(|_| internal_server_error(ErrorMessages::INTERNAL_SERVER, None))?;

            if target_account.is_none() {
                return Err(not_found_error(ErrorMessages::ACCOUNT_NOT_FOUND_TARGET, None));
            }
        }

        if (tipo == TransactionType::Transfer || tipo == TransactionType::Withdrawal) 
            && origin_account.is_some() 
            && origin_account.as_ref().unwrap().saldo < Some(BigDecimal::from(data.valor as i64))
        {
            return Err(bad_request_error(ErrorMessages::ACCOUNT_BAD_REQUEST_BALANCE_NOT_ENOUGH, None));
        }

        // Iniciar transação
        let mut tx: sqlx::Transaction<'_, sqlx::Postgres> = database::db().begin()
            .await
            .map_err(|_| internal_server_error(ErrorMessages::INTERNAL_SERVER, None))?;

        match tipo {
            TransactionType::Deposit => {
                AccountRepository::add_balance(&data.id_conta_destino.unwrap().to_string(), data.valor, &mut tx)
                    .await
                    .map_err(|_| internal_server_error(ErrorMessages::INTERNAL_SERVER, None))?;
            }
            TransactionType::Transfer => {
                AccountRepository::remove_balance(&data.id_conta_origem.unwrap().to_string(), data.valor, &mut tx)
                    .await
                    .map_err(|_| internal_server_error(ErrorMessages::INTERNAL_SERVER, None))?;
                AccountRepository::add_balance(&data.id_conta_destino.unwrap().to_string(), data.valor, &mut tx)
                    .await
                    .map_err(|_| internal_server_error(ErrorMessages::INTERNAL_SERVER, None))?;
            }
            TransactionType::Withdrawal => {
                AccountRepository::remove_balance(&data.id_conta_origem.unwrap().to_string(), data.valor, &mut tx)
                    .await
                    .map_err(|_| internal_server_error(ErrorMessages::INTERNAL_SERVER, None))?;
            }
        };


        let new_transaction = TransactionRepository::create(
            ICreateTransactionData {
                tipo: tipo.to_string(),
                valor: data.valor,
                id_conta_origem: data.id_conta_origem,
                id_conta_destino: data.id_conta_destino,
            },
            &mut tx
        ).await
        .map_err(|_| internal_server_error(ErrorMessages::INTERNAL_SERVER, None))?;

        tx.commit()
            .await
            .map(|_| new_transaction)
            .map_err(|_| internal_server_error(ErrorMessages::INTERNAL_SERVER, None))
    }

    pub async fn find_all(
        id_account: Uuid,
        id_client: String,
        role: RoleEnum,
    ) -> Result<Vec<TransactionEntity>, BaseError> {
        let permission = permission::has_permission(role, RoleEnum::Manager);

        let account = AccountRepository::find_by_id(&id_account.to_string(), false)
            .await
            .map_err(|_| internal_server_error(ErrorMessages::INTERNAL_SERVER, None))?;

        if account.is_none() {
            return Err(not_found_error(ErrorMessages::ACCOUNT_NOT_FOUND, None));
        }

        let account = account.unwrap();
        if account.id_cliente.unwrap().to_string() != id_client && !permission {
            return Err(unauthorized_error(ErrorMessages::UNAUTHORIZED, None));
        }

        let transactions = TransactionRepository::find_all(&id_account.to_string())
            .await
            .map_err(|_| internal_server_error(ErrorMessages::INTERNAL_SERVER, None))?;

        if transactions.is_empty() {
            return Ok(Vec::new());
        }

        Ok(transactions)
    }
} 